### PR TITLE
docs(NODE-3753): Update testing readme file

### PR DESCRIPTION
**Pull request by ljhaywar**
### Description

[NODE 3753](https://jira.mongodb.org/browse/NODE-3753)

#### What is changing?

Updating the Test readme to have more detailed information about the tests

##### Is there new documentation needed for these changes?

No

#### What is the motivation for this change?

The readme was lacking details that newbies need to understand the test infrastructure

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the correct format: `<type>(NODE-xxxx)<!>: <description>`
- [ ] ~~Changes are covered by tests~~
- [ ] ~~New TODOs have a related JIRA ticket~~

### CODE REVIEW

**Comment by ljhaywar** [comment]
Some notes...

Benchmark tests explanation - this was very much an educated guess. Is there any more info on those that we can point people to?

Using a Pre-Release Version of a Dependent Library. I based the instructions on Daria’s comments in the Jira ticket. I was able to follow the steps, and the tests didn’t blow up, but I wasn’t sure how to verify I was actually using the pre-release version.

I left TODOs for explaining what the environment variables represent and how to get their values. If someone wants to take a crack at explaining that, I’m happy to clean up the wording.

I left “TODO Special Env Sections” alone. If someone wants to take a crack at those, I’m happy to clean up the wording.

File: test/readme.md [hunk 1]
```diff
@@ -1,110 +1,268 @@
...
+| Benchmark | `/test/benchmarks` | The benchmark tests report how long a designated set of tests take to run. They are used to measure performance. | `npm run check:bench` |
+| Integration | `/test/integration` | *Coming Soon* The integration tests test that pieces of the driver work together as expected. | `npm run check:test` |
+| Specialized Environment | `/test/manual` | The specalized environment tests are functional tests that require specialized environment setups in Evergreen. <br><br>**Note**: "manual" in the directory path does not refer to tests that should be run manually. These tests are automated. These tests require manual configuration in Evergreen. | There is no single script for running all of the specialized environment tests. Instead, you can run the appropriate script based on the specialized environment you want to use: <br>- `npm run check:atlas` to test Atlas <br>- `npm run check:adl` to test Atlas Data Lake <br>- `npm run check:ocsp` to test OSCP <br>- `npm run check:kerberos` to test Kerberos <br>- `npm run check:tls` to test TLS <br>- `npm run check:ldap` to test LDAP authorization
```
Comment (durran) [inline test/readme.md:31]:
```suggestion
| Specialized Environment | `/test/manual` | The specalized environment tests are functional tests that require specialized environment setups in Evergreen. <br><br>**Note**: "manual" in the directory path does not refer to tests that should be run manually. These tests are automated. These tests have a special evergreen configuration and run in isolation from the other tests.. | There is no single script for running all of the specialized environment tests. Instead, you can run the appropriate script based on the specialized environment you want to use: <br>- `npm run check:atlas` to test Atlas <br>- `npm run check:adl` to test Atlas Data Lake <br>- `npm run check:ocsp` to test OSCP <br>- `npm run check:kerberos` to test Kerberos <br>- `npm run check:tls` to test TLS <br>- `npm run check:ldap` to test LDAP authorization
```

File: test/readme.md [hunk 2]
```diff
@@ -1,110 +1,268 @@
...
+| Spec | Test input and expected results are in `/test/spec`.  <br>Test runners are in `/test/functional` with  the `_spec` suffix in the test file's name.  <br>Some spec tests are also in `/test/unit`. | All of the MongoDB drivers follow the same [specifications (specs)][driver-specs]. Each spec has tests associated with it. Some of the tests are prose (written, descriptive) tests.  Other tests are written in JSON. The developers on the driver teams automate these tests. For prose tests, the tests are converted to automation and stored in the `test/unit` or `test/functional` directory as appropriate. For the JSON tests, a developer uses a tool to convert the JSON test file to YAML. Both the JSON and YAML files are stored in `test/spec`. The test runners for the JSON and YAML files are located in `test/functional` and `/test/unit`. | `npm run check:test` to run all of the functional and integration tests (including the spec tests stored with those). `npm run check:unit` to run all of the unit tests (including the spec tests stored with those).
```
Comment (durran) [inline test/readme.md:32]:
```suggestion
| Spec | Test input and expected results are in `/test/spec`.  <br>Test runners are in `/test/functional` with  the `_spec` suffix in the test file's name.  <br>Some spec tests are also in `/test/unit`. | All of the MongoDB drivers follow the same [specifications (specs)][driver-specs]. Each spec has tests associated with it. Some of the tests are prose (written, descriptive) tests.  Other tests are written in YAML and converted to JSON. The developers on the driver teams automate these tests. For prose tests, the tests are converted to automation and stored in the `test/unit` or `test/functional` directory as appropriate. Both the JSON and YAML files are stored in `test/spec`. The test runners for the JSON and YAML files are located in `test/functional` and `/test/unit`. | `npm run check:test` to run all of the functional and integration tests (including the spec tests stored with those). `npm run check:unit` to run all of the unit tests (including the spec tests stored with those).
```

File: test/readme.md [hunk 3]
```diff
@@ -1,110 +1,268 @@
...
+
+In the steps above, we started a standalone server: `./test/tools/cluster_setup.sh server`.
+
+You can use the same [cluster_setup.sh](tools/cluster_setup.sh) script to start a replica set or sharded cluster by passing the appropriate option: `./test/tools/cluster_setup.sh replica_set` or
+`./test/tools/cluster_setup.sh sharded_cluster`.  If you are running more than a standalone server, make sure your `ulimit` settings are in accordance with [MongoDB's recommendations][mongodb-ulimit]. Changing the settings on the latest versions of macOS can be tricky. See [this article][macos-ulimt] for tips. (You likely don't need to do the complicated maxproc steps.)
+
+The [cluster_setup.sh](tools/cluster_setup.sh) script automatically stores the files associated with the MongoDB server in the `data` directory, which is stored at the top level of this repository.
+You can delete this directory if you want to ensure you're running a clean configuration. If you delete the directory, the associated database server will be stopped, and you will need to run [cluster_setup.sh](tools/cluster_setup.sh) again.
```
Comment (durran) [inline test/readme.md:60]: I'm actually not sure if this is true - I think if the server is running you wouldn't be allowed to delete the directory as the server would have a handle on those files. I think you would need to stop the server first, then delete the directory.

File: test/readme.md [hunk 4]
```diff
@@ -1,110 +1,268 @@
...
+
+You can prefix `npm test` with a `MONGODB_URI` environment variable to point the tests to a specific deployment. For example, for a standalone server, you might use: `env MONGODB_URI=mongodb://localhost:27017 npm test`. For a replica set, you might use: `env MONGODB_URI=mongodb://localhost:31000,localhost:31001,localhost:31002/?replicaSet=rs npm test`.
```
Comment (durran) [inline test/readme.md:62]: `env` is not necessary here in the command.

File: test/readme.md [hunk 5]
```diff
@@ -1,110 +1,268 @@
...
+
+### Running Individual Tests
+
+The easiest way to run a single test is by appending `.only()` to the suite or test you want to run. For example, you could update a test function to be `it.only(‘cool test’, function() {})`.  Then
```
Comment (durran) [inline test/readme.md:66]:
```suggestion
The easiest way to run a single test is by appending `.only()` to the test context you want to run. For example, you could update a test function to be `it.only(‘cool test’, function() {})`.  Then
```

File: test/readme.md [hunk 6]
```diff
@@ -1,110 +1,268 @@
...
+run the test using `npm run check:test` for a functional or integration test or `npm run check:unit` for a unit test.  See [Mocha's documentation][mocha-only] for more detailed information on `.only()`.
+
+Another way to run a single test is to use Mocha's `grep` flag.  For functional or integration tests, run `npm run check:test -- -g 'test name'`. For unit tests, run `npm run check:unit -- -g 'test name'`. See the [Mocha documentation][mocha-grep] for information on the `grep` flag.
+
+## Running the Tests in Evergreen
+
+[Evergreen][evergreen-wiki] is the continuous integration (CI) system we use.  Evergreen builds are automatically run whenever a pull request is created or when commits are pushed to particular branches (e.g., main, 4.0, and 3.6).
+
+Each Evergreen build runs the test suite against a variety of build variants that include a combination of topologies, special environments, and operating systems. By default, commits in pull requests only run a subset of the build variants in order to save time and resources. These builds can be individually configured in the Evergreen UI to include more build variants.
```
Comment (durran) [inline test/readme.md:75]:
```suggestion
Each Evergreen build runs the test suite against a variety of build variants that include a combination of topologies, special environments, and operating systems. By default, commits in pull requests only run a subset of the build variants in order to save time and resources. These builds can be individually configured in the `.evergreen/config.yml.in` and then by generating a new Evergreen config via `node .evergreen/generate_evergreen_tasks.js`.
```

File: test/readme.md [hunk 7]
```diff
@@ -1,110 +1,268 @@
...
+## Using a Pre-Release Version of a Dependent Library
 
-```sh
-FAKE_MONGODB_SERVICE_ID="true"
-```
+You may want to test the driver with a pre-release version of a dependent library (e.g., [bson][js-bson]).
+Follow the steps below to do so.
+1. Open [package.json](../package.json)
+1. Identify the line that specifies the dependency
+1. Replace the version number with the commit hash of the dependent library. For example: `"bson": "e29156f7438fa77c1672fd70789d7ade9ca65061"`
```
Comment (durran) [inline test/readme.md:115]:
```suggestion
1. Replace the version number with the commit hash of the dependent library. For example, js-bson project on Github: `"bson": "mongodb/js-bson#e29156f7438fa77c1672fd70789d7ade9ca65061"`
```

File: test/readme.md [hunk 8]
```diff
@@ -1,110 +1,268 @@
...
-Take note of the `[ topology type: load-balanced ]` printout from mocha to make sure it picked up the environment as expected.
+1. Navigate to a new directory and create a new Node project by running `npm init` in a terminal and    working through the interactive prompts. A new file named `package.json` will be created for you.
+1. In `package.json`, create a new dependency for `mongodb` that points to your local copy of the driver. For example:
+   ```
+   "dependencies": {
+     "mongodb": "/path-to-your-copy-of-the-driver-repo/node-mongodb-native"
+   }
+   ```
+1. Run `npm install` to install the dependency.
+1. Create a new file that uses the driver to test your changes. See the [MongoDB Node.js Quick Start Repo][node-quick-start] for example scripts you can use.
```
Comment (durran) [inline test/readme.md:134]: I'd add something like, "Note when making driver changes you will have to run `npm run build:ts` with each change to have it take effect".

File: test/readme.md [hunk 9]
```diff
@@ -1,110 +1,268 @@
...
+
+1. Create the following environment variables using a command like `export PROJECT="node-driver"`.
+
+   - `PROJECT`
+   - `SERVERLESS_DRIVERS_GROUP`
+   - `SERVERLESS_API_PUBLIC_KEY`
+   - `SERVERLESS_API_PRIVATE_KEY`
+   - `SERVERLESS_ATLAS_USER`
+   - `SERVERLESS_ATLAS_PASSWORD`
+   - `LOADBALANCED`
```
Comment (durran) [inline test/readme.md:159]: Did you find this environment variable is needed? I didn't see where it was used in the scripts.

File: test/readme.md [hunk 10]
```diff
@@ -1,110 +1,268 @@
...
+
+   TODO:  Explain what these variables represent and how to get their values.
```
Comment (durran) [inline test/readme.md:161]: Maybe add: "If you are an Atlas user the values can be found in your project details. If you are testing for driver work, consult the Evergreen project configuration for these values."

File: test/readme.md [hunk 11]
```diff
@@ -1,110 +1,268 @@
...
+
+1. Create the following environment variables using a command like `export AWS_REGION="us-east-1"`.
+
+   - `AWS_ACCESS_KEY_ID`
+   - `AWS_SECRET_ACCESS_KEY`
+   - `CSFLE_KMS_PROVIDERS`
+   - `AWS_REGION`
+   - `AWS_CMK_ID`
+
+   TODO:  Explain what these variables represent and how to get their values.
```
Comment (durran) [inline test/readme.md:258]: Make a note for drivers testing to pull from Evergreen, otherwise pull them from the AWS account.

**Review by durran — CHANGES_REQUESTED** [review]

File: test/readme.md (hunk 3 above)
Comment (ljhaywar) [reply test/readme.md:60]: This is taken from one of the comments Daria added to https://jira.mongodb.org/browse/NODE-3753. I've tested it out, and the processes are no longer running after deleting the directory. I'm happy to change it if you don't think people should take this approach.

File: test/readme.md (hunk 4 above)
Comment (ljhaywar) [reply test/readme.md:62]: Thank you!

File: test/readme.md (hunk 8 above)
Comment (ljhaywar) [reply test/readme.md:134]: Great catch! Thank you!

File: test/readme.md (hunk 3 above)
Comment (durran) [reply test/readme.md:60]: I just verified that as well, so fine to stay as is.

File: test/readme.md (hunk 9 above)
Comment (ljhaywar) [reply test/readme.md:159]: For traceability, we tried this out together and decided it wasn't needed.

File: test/readme.md [hunk 12]
```diff
@@ -1,110 +1,268 @@
...
+The following steps will walk you through how to run the tests for CSFLE.
+
+1. Install [MongoDB Client Encryption][npm-csfle] if you haven't already:
+   `npm install mongodb-client-encryption`
+
+1. Create the following environment variables using a command like `export AWS_REGION="us-east-1"`.
+
+   - `AWS_ACCESS_KEY_ID`
+   - `AWS_SECRET_ACCESS_KEY`
+   - `CSFLE_KMS_PROVIDERS`
```
Comment (durran) [inline test/readme.md:254]: I'd describe CSFLE_KMS_PROVIDERS as "The raw EJSON description of the KMS providers". The format is:

```
interface CSFLE_kms_providers {
   aws: {
     accessKeyId: string;
     secretAccessKey: string;
  };
  azure: {
    tenantId: string;
    clientId: string;
    clientSecret: string;
  };
  gcp: {
    email: string;
    privateKey: string;
  };
  local: {
    // EJSON handle converting this, its actually the canonical -> { $binary: { base64: string; subType: string } }
    // **NOTE**: The dollar sign has to be escaped when using this as an ENV variable
    key: Binary;
  }
}
```

Comment (ljhaywar) [reply test/readme.md:254]: Than you for this!

**Comment by ljhaywar** [comment]
@durran Thanks for the feedback! I've made updates. Can you take another look at your convenience?

**Review by durran — APPROVED** [review]

**Review by nbbeeken — APPROVED** [review]
Thanks so much! this is a great write up 😃

**Review by dariakp — APPROVED** [review]
This is great, thank you so much for the work you did here!

**Review by baileympearson — APPROVED** [review]
The changes look good to me (given my limited experience) and they've already proved helpful.  Thanks for taking this on!  😃

**Pull request merged**